File: Practices/pigeon.js
function hasSharedBirthday(n) {
    let birthday = Array(366).fill(false);

    for (let i = 0; i < n; i++) {
        let bday = Math.ceil(Math.random() * 365);
        if (birthday[bday]) return true;
        birthday[bday] = true;
    }
    return false;
}

function simulate(groupSize, trials) {
    let sharedCount = 0;
    for (let i = 0; i < trials; i++) {
        if (hasSharedBirthday(groupSize)) sharedCount++;
    }
    console.log(`Probability of shared birthday in group of ${groupSize}: ${(sharedCount / trials * 100).toFixed(2)}%`);
}

simulate(23, 10000);

File: Practices/pigeon.py
import random

def has_shared_birthday(n):
    birthdays = set()

    for _ in range(n):
        # Generates a random day between 1 and 365 inclusive
        bday = random.randint(1, 365) 
        if bday in birthdays:
            return True
        birthdays.add(bday)
        
    return False

def simulate(group_size, trials):
    shared_count = 0
    for _ in range(trials):
        if has_shared_birthday(group_size):
            shared_count += 1
            
    probability = (shared_count / trials) * 100
    print(f"Probability of shared birthday in group of {group_size}: {probability:.2f}%")

# Run the simulation
simulate(23, 10000)